launchable/commands/record/commit.py

```diff
@@ -10,11 +19,67 @@
               default=os.getcwd(),
               type=str
               )
-def commit(source):
-    token, org, workspace = parse_token()
+@click.option('--executable',
+              help="collect commits with Jar or Docker",
+              type=click.Choice(['jar', 'docker']),
+              default='jar'
+              )
+def commit(source, executable):
     source = os.path.abspath(source)
+
+    if executable == 'jar':
+        download_jar()
+        exec_jar(source)
+    elif executable == 'docker':
+        exec_docker(source)
+
+
+def download_jar():
+    pathlib.Path(jar_dir_path).mkdir(parents=True, exist_ok=True)
+
+    res = requests.get(ingester_jar_url, stream=True)
+    total_length = int(res.headers.get('content-length'))
+
+    if os.path.exists(jar_file_path) and os.path.getsize(jar_file_path) == total_length:
+        print("exe_deploy.jar exists in {}".format(jar_dir_path))
+    else:
+        total_downloaded = 0
+        f = open(jar_file_path, 'wb')
+        for chunk in res.iter_content(chunk_size=512 * 1024):
+            if chunk:
+                f.write(chunk)
+
+                total_downloaded += len(chunk)
+                sys.stdout.write("\rDownloading exe_deploy.jar: {:.1f}% ({:,} / {:,} byte)".format(
+                    (total_downloaded / total_length) * 100,
+                    total_downloaded,
+                    total_length
+                ))
+                sys.stdout.flush()
+
+        f.close()
+        sys.stdout.write("\nexe_deploy.jar is downloaded in {}\n".format(jar_dir_path))
+
+
+def exec_jar(source):
+    res = subprocess.run(["which", "java"], stdout=subprocess.DEVNULL)
+    java = None
+    if res.returncode == 0:
+        java = "java"
+    elif os.access(os.path.expandvars("$JAVA_HOME/bin/java"), os.X_OK):
+        java = "$JAVA_HOME/bin/java"
+
+    if not java:
+        exit("You need to install Java or try --executable docker")
+
+    os.system(
```

Comment:
When the commit ingester fails, how should we handle it? It seems to me that the code today will silently ignore that and moves on. Perhaps that's intended (so as not to disrupt the customer's build/test), or maybe it's an oversight.

More broadly, this does seem like something we should think carefully. There are competing factors we need to balance -- we don't want to disrupt customers' build & test and get kicked out of their build process, but we also want them to notice if things aren't working as it's supposed to be.

Whatever our policy is that we define there, we should apply that consistently in other commands, too.

Reply:
It's simply an oversight. I think `LAUNCHABLE_REPORT_ERROR` environment variable is a good option for this fail.  I'll fix it at another PR after merge this PR.